### PR TITLE
Move closer to upstream codebase.

### PR DESCRIPTION
**Pull request by Xazax-hun**
I compared the upstream version of the ASTImporter to our version and eliminated some differences. This should bring us closer to upstream. It might fix some bugs, there are some code style improvements and also it will make it easier to port our changes to clang 6 in the future.
It also uncovered some differences that might need some investigation. Also, some tests were missing.

### CODE REVIEW

File: lib/AST/ASTImporter.cpp [hunk 1]
```diff
@@ -4595,22 +4650,20 @@ Decl *ASTNodeImporter::VisitFunctionTemplateDecl(FunctionTemplateDecl *D) {
...
   }
 
-  TemplateParameterList *Params = ImportTemplateParameterList(
-                                    D->getTemplateParameters());
+  TemplateParameterList *Params =
+      ImportTemplateParameterList(D->getTemplateParameters());
   if (!Params)
     return nullptr;
 
+  FunctionDecl *TemplatedFD =
```
Comment (Xazax-hun) [inline lib/AST/ASTImporter.cpp]: This code is moved down from above. It make more sense to try to import the templated function only after we made sure no such function already exist.

File: lib/AST/ASTImporter.cpp [hunk 2]
```diff
@@ -2542,6 +2554,7 @@ Decl *ASTNodeImporter::VisitFunctionDecl(FunctionDecl *D) {
   }
 
   // Import the describing template function, if any.
+  // FIXME: This is omitted in upstream code. Do we actually need this?
```
Comment (Xazax-hun) [inline lib/AST/ASTImporter.cpp]: @szepet , IIRC you were dealing with template functions. What do you think about this FIXME?

File: lib/AST/ASTImporter.cpp [hunk 3]
```diff
@@ -4337,7 +4357,7 @@ Decl *ASTNodeImporter::VisitClassTemplateSpecializationDecl(
 
     D2->setTemplateSpecializationKind(D->getTemplateSpecializationKind());
 
-    // Set the context of this specialization/instantiation.
+    // Add the specialization to this context.
```
Comment (martong) [inline lib/AST/ASTImporter.cpp]: Actually here, the original comment (the removed one) is the true one, since we don't add anything to the declcontext.

**Comment by bruntib** [comment]
Are we going to diff our version with the original `ASTImporter.cpp` in the long terms? There are still some differences. However this requires more effort to check each difference if it was us or the upstream who modified it.

**Comment by Xazax-hun** [comment]
I think yes, we always want to ship the latest and greatest version of the importer. There are some remaining diffs, but apart from the one I marked with FIXME all of them are changes that we did not upstream yet.

**Comment by dkrupp** [comment]
Check with @szepet FIXME in ASTImporter.cpp 2557 then  Run before/after on curl, xerces, postgesql to see if there is any regression.

**Comment by Xazax-hun** [comment]
After some discussion with @szepet I removed the FIXME. I think this should be committed as is, since most of the changes are formatting related and additional tests. This should be safe but keeping it around without merging would increase the conflicts.

**Pull request merged**